pkg/controller/master/template/template_controller.go

```diff
@@ -0,0 +1,33 @@
+package template
+
+import (
+	"reflect"
+
+	apisv1alpha1 "github.com/rancher/harvester/pkg/apis/vm.cattle.io/v1alpha1"
+	ctlapisv1alpha1 "github.com/rancher/harvester/pkg/generated/controllers/vm.cattle.io/v1alpha1"
+	"github.com/rancher/harvester/pkg/ref"
+)
+
+type templateHandler struct {
+	templates            ctlapisv1alpha1.TemplateClient
+	templateVersionCache ctlapisv1alpha1.TemplateVersionCache
+}
+
+func (h *templateHandler) OnChanged(key string, tp *apisv1alpha1.Template) (*apisv1alpha1.Template, error) {
+	if tp.Spec.DefaultVersionID == "" {
+		return tp, nil
+	}
+
+	versionNs, versionName := ref.Parse(tp.Spec.DefaultVersionID)
+	version, err := h.templateVersionCache.Get(versionNs, versionName)
+	if err != nil {
+		return nil, err
+	}
+
+	copyTp := tp.DeepCopy()
+	copyTp.Status.DefaultVersion = version.Status.Version
+	if !reflect.DeepEqual(copyTp, tp) {
+		return h.templates.Update(copyTp)
+	}
```

Comment:
We can compare the version directly. Only do deep copy when an update is needed and avoid the deepEqual check.